luigi/contrib/hdfs/hadoopcli_clients.py

```diff
@@ -0,0 +1,288 @@
+        else:
+            cmd = load_hadoop_cmd() + ['fs', '-getmerge', path, local_destination]
+        self.call_check(cmd)
+
+    def mkdir(self, path, parents=True, raise_if_exists=False):
+        if (parents and raise_if_exists):
+            raise NotImplementedError("HdfsClient.mkdir can't raise with -p")
+        try:
+            cmd = (load_hadoop_cmd() + ['fs', '-mkdir'] +
+                   (['-p'] if parents else []) +
+                   [path])
+            self.call_check(cmd)
+        except hdfs_error.HDFSCliError as ex:
+            if "File exists" in ex.stderr:
+                if raise_if_exists:
+                    raise FileAlreadyExists(ex.stderr)
+            else:
+                raise
+
+    def listdir(self, path, ignore_directories=False, ignore_files=False,
+                include_size=False, include_type=False, include_time=False, recursive=False):
+        if not path:
+            path = "."  # default to current/home catalog
+
+        if recursive:
+            cmd = load_hadoop_cmd() + ['fs'] + self.recursive_listdir_cmd + [path]
+        else:
+            cmd = load_hadoop_cmd() + ['fs', '-ls', path]
+        lines = self.call_check(cmd).split('\n')
+
+        for line in lines:
+            if not line:
+                continue
+            elif line.startswith('OpenJDK 64-Bit Server VM warning') or line.startswith('It\'s highly recommended') or line.startswith('Found'):
+                continue  # "hadoop fs -ls" outputs "Found %d items" as its first line
+            elif ignore_directories and line[0] == 'd':
+                continue
+            elif ignore_files and line[0] == '-':
+                continue
+            data = line.split(' ')
+
+            file = data[-1]
+            size = int(data[-4])
+            line_type = line[0]
+            extra_data = ()
+
+            if include_size:
+                extra_data += (size,)
+            if include_type:
+                extra_data += (line_type,)
+            if include_time:
+                time_str = '%sT%s' % (data[-3], data[-2])
+                modification_time = datetime.datetime.strptime(time_str,
+                                                               '%Y-%m-%dT%H:%M')
+                extra_data += (modification_time,)
+
+            if len(extra_data) > 0:
+                yield (file,) + extra_data
+            else:
+                yield file
+
+    def touchz(self, path):
+        self.call_check(load_hadoop_cmd() + ['fs', '-touchz', path])
+
+
+class HdfsClientCdh3(HdfsClient):
+    """
+    This client uses CDH3 syntax for file system commands.
+    """
+
+    def mkdir(self, path):
+        """
+        No -p switch, so this will fail creating ancestors.
+        """
+        try:
+            self.call_check(load_hadoop_cmd() + ['fs', '-mkdir', path])
+        except hdfs_error.HDFSCliError as ex:
+            if "File exists" in ex.stderr:
```

Comment:
I'm not familiar with the codebase, but it seems like the exception handling logic could be moved to a private method to avoid code duplication. See line 188 for example.